test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs

```diff
@@ -1413,6 +1413,72 @@ public async Task DotnetRestore_MultiTargettingProject_WithDifferentProjectRefer
+                // Library project with Framework Reference
+                var libraryName = "Library";
+                var libraryProjectFilePath = Path.Combine(Path.Combine(pathContext.SolutionRoot, libraryName), $"{libraryName}.csproj");
+                _msbuildFixture.CreateDotnetNewProject(pathContext.SolutionRoot, libraryName, "classlib");
+
+                using (var stream = File.Open(libraryProjectFilePath, FileMode.Open, FileAccess.ReadWrite))
+                {
+                    var xml = XDocument.Load(stream);
+                    ProjectFileUtils.AddItem(
+                        xml,
+                        name: "FrameworkReference",
+                        identity: "Microsoft.AspNetCore.App",
+                        framework: (string)null,
+                        attributes: new Dictionary<string, string>(),
+                        properties: new Dictionary<string, string>()
+                        );
+                    ProjectFileUtils.WriteXmlToFile(xml, stream);
+                }
+
+                var packResult = _msbuildFixture.RunDotnet(pathContext.SolutionRoot, $"pack {libraryProjectFilePath} /p:PackageOutputPath=\"{pathContext.PackageSource}\"", ignoreExitCode: true);
+                packResult.Success.Should().BeTrue(because: packResult.AllOutput);
+
+                // Consumer project.
+                var consumerName = "Consumer";
+                _msbuildFixture.CreateDotnetNewProject(pathContext.SolutionRoot, consumerName, "console");
+
+                var consumerProjectFilePath = Path.Combine(Path.Combine(pathContext.SolutionRoot, consumerName), $"{consumerName}.csproj");
+
+                using (var stream = File.Open(consumerProjectFilePath, FileMode.Open, FileAccess.ReadWrite))
+                {
+                    var xml = XDocument.Load(stream);
+                    ProjectFileUtils.AddItem(
+                        xml,
+                        name: "PackageReference",
+                        identity: "Library",
+                        framework: (string)null,
+                        attributes: new Dictionary<string, string>() { { "Version", "1.0.0" } },
+                        properties: new Dictionary<string, string>()
+                        );
+                    ProjectFileUtils.WriteXmlToFile(xml, stream);
+                }
+
+                // Add a DBT.
+
+                var directoryBuildTargetsFilePath = Path.Combine(pathContext.SolutionRoot, "Directory.Build.targets");
+                File.WriteAllText(directoryBuildTargetsFilePath,
+@"<Project>
+    <Target Name=""PrintFrameworkReferences"" AfterTargets=""Build"" DependsOnTargets=""CollectFrameworkReferences"">
+        <Message Text=""Framework References: '@(_FrameworkReferenceForRestore)'"" Importance=""High"" />
+    </Target>
+</Project>");
+
+                var buildResult = _msbuildFixture.RunDotnet(pathContext.SolutionRoot, $"build {consumerProjectFilePath}", ignoreExitCode: true);
```

Comment:
Would these types of tests be better off using the MSBuild API to load the project, then just check all `_FrameworkReferenceForRestore` items directly, rather than invoking a new process, capturing the output, then checking all lines in the output?

Reply:
Not really. 
Various reasons:

* The cleanup logic is happening inside a target. You'd need to call the target. 
* The problem happens only when the build targets are running alongside the restore collection targets. The build targets use the info from the previous restore to generate extra framework references.